### PR TITLE
[Draft] just-in-time (JIT) meta computations

### PR DESCRIPTION
**Pull request by dshemetov**
Addresses a part of #646.

Contains work for computing meta values for the JIT signals. The basic setup is this:
- base signals (non-derived, aka non-JIT, signals) have their meta computed normally
- derived signals make a request to a local API server and then use Pandas to compute the resulting meta values

This work refactors `database.py` and moves the meta computations into `database_meta.py`. The main work is in the latter file. The main testing is in `test_covidcast_meta_cache_updater.py`.

Would appreciate reviews on testing strategy, testing completeness, and clarity of the code.

TODO:
- [ ] Make complete tests.
  - [ ] Make a table of the integration tests that already exist for JIT computations.
    - [ ] List all the queries that have been tested. List all the possible edge cases.
  - [ ] Make a table of the meta computation A/B tests.
    - [ ] List all the queries that have been tested. List all the possible edge cases.

### CODE REVIEW

**Comment by dshemetov** [comment]
### Exposition on A/B Tests

First, I took a chunk of data from `jhu-csse:confirmed_cumulative_num` between 2022-04-01 and 2022-06-01 to test my meta computations. I chose 4 counties and 2 states.

I ran an A/B test between my meta just-in-time (JIT) computations and the existing database approach. All the values matched except for the `max_lag` field. Not only that, but *all* the values matched for `confirmed_cumulative_num`, the derived signals didn't though.

I was at a loss. How can the derived values possibly have different lags?

Katie figured it out immediately: the update profiles look completely different for cumulative versus incidence. Consider: if we backfill 1 count to a cumulative signal to a day in the past, then all the consecutive days need to be updated (since the entire curve from that day on is raised by 1). For the incidence signal, however, this will just be a single backfilled blip and the rest of the values are unaffected. So if the cumulative signal is reissued, its lag/issue field is very likely to fall out of sync with the incidence signal.

![[Pasted image 20220721131752.png]]

*This plot shows lag against time for 3 signals, for a single county.*

However, when computing the diffed signal JIT, we only have access to a rolling window of two rows, from which we must decide on the issue (we can't look arbitrarily far in the past to see when `confirmed_cumulative_num` was updated, for that defeat the server's streaming architecture). So JIT has no alternative but to deviate here from existing functionality and choose an issue from one of the two rows it has access to when diffing.

But more on that later.

To see if I could avoid this issue, I instead opted to collect data from the very beginning of the `jhu-csse` signal history. The earliest data available is on 2020-01-22, so I grabbed data from then until 2020-05-01. It turned out that there were data discrepancies there too. **This results in `test-data1.csv`**.

Consider the following query for a single county for the earliest data possible.

```python
>>> start_day = date(2020, 1, 15)
>>> end_day = date(2020, 1, 23)
>>> df1 = covidcast.signal(data_source="jhu-csse", signal="confirmed_cumulative_num", start_day=start_day, end_day=end_day, time_type="day", geo_type="county", geo_values="02100")
>>> df2 = covidcast.signal(data_source="jhu-csse", signal="confirmed_incidence_num", start_day=start_day, end_day=end_day, time_type="day", geo_type="county", geo_values="02100")
>>> >>> df3 = covidcast.signal(data_source="jhu-csse", signal="confirmed_7dav_incidence_num", start_day=start_day, end_day=end_day + timedelta(days=30), time_type="day", geo_type="county", geo_values="02100")
>>> d1
	geo_value	signal	time_value	issue	lag	missing_value	missing_stderr	missing_sample_size	value	stderr	sample_size	geo_type	data_source
0	02100	confirmed_cumulative_num	2020-01-22	2020-05-14	113	0	5	5	0.0	None	None	county	jhu-csse
0	02100	confirmed_cumulative_num	2020-01-23	2020-05-14	112	0	5	5	0.0	None	None	county	jhu-csse

>>> df2
geo_value	signal	time_value	issue	lag	missing_value	missing_stderr	missing_sample_size	value	stderr	sample_size	geo_type	data_source
0	02100	confirmed_incidence_num	2020-01-22	2020-05-14	113	0	5	5	0.0	None	None	county	jhu-csse
0	02100	confirmed_incidence_num	2020-01-23	2020-05-14	112	0	5	5	0.0	None	None	county	jhu-csse

>>> df3
	geo_value	signal	time_value	issue	lag	missing_value	missing_stderr	missing_sample_size	value	stderr	sample_size	geo_type	data_source
0	02100	confirmed_7dav_incidence_num	2020-02-20	2021-04-01	406	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_7dav_incidence_num	2020-02-21	2021-04-01	405	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_7dav_incidence_num	2020-02-22	2021-04-01	404	0	0	0	0.0	None	None	county	jhu-csse
```

- discrepancy 1: the earliest issues available for incidence match cumulative. While mostly inconsequential and the earliest incidence value is filled with a 0, we couldn't actually have access to this data, so it must be a spoof
- discrepancy 2: the earliest smoothed data available is on 2020-02-20, which lags behind the earliest cumulative data by over a month

So for that reason, I decided to bump my data up a little later, so that the earliest data available for cumulative and incidence matches the smoothed signal (so I chose cumulative 2020-02-13 -- 2020-05-01 and incidence 2020-02-14 -- 2020-05-01). **This results in `test-data2.csv`**.

This data has issues too.

I started with having discrepancies between the `min_lag` values in my JIT code and in the data. Here's what the data showed:

```python
>>> end_day = date(2020, 5, 1)
>>> df1 = covidcast.signal(data_source="jhu-csse", signal="confirmed_cumulative_num", start_day=end_day - timedelta(days=7), end_day=end_day, time_type="day", geo_type="county", geo_values="02100")
>>> df2 = covidcast.signal(data_source="jhu-csse", signal="confirmed_incidence_num", start_day=end_day - timedelta(days=6), end_day=end_day, time_type="day", geo_type="county", geo_values="02100")
>>> df3 = covidcast.signal(data_source="jhu-csse", signal="confirmed_7dav_incidence_num", start_day=end_day, end_day=end_day, time_type="day", geo_type="county", geo_values="02100")
>>> df1
geo_value	signal	time_value	issue	lag	missing_value	missing_stderr	missing_sample_size	value	stderr	sample_size	geo_type	data_source
0	02100	confirmed_cumulative_num	2020-04-24	2021-04-01	342	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_cumulative_num	2020-04-25	2021-04-01	341	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_cumulative_num	2020-04-26	2021-04-01	340	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_cumulative_num	2020-04-27	2021-04-01	339	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_cumulative_num	2020-04-28	2021-04-01	338	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_cumulative_num	2020-04-29	2021-04-01	337	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_cumulative_num	2020-04-30	2021-04-01	336	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_cumulative_num	2020-05-01	2021-04-01	335	0	0	0	0.0	None	None	county	jhu-csse
>>> df2
geo_value	signal	time_value	issue	lag	missing_value	missing_stderr	missing_sample_size	value	stderr	sample_size	geo_type	data_source
0	02100	confirmed_incidence_num	2020-04-25	2021-04-01	341	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_incidence_num	2020-04-26	2021-04-01	340	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_incidence_num	2020-04-27	2021-04-01	339	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_incidence_num	2020-04-28	2021-04-01	338	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_incidence_num	2020-04-29	2021-04-01	337	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_incidence_num	2020-04-30	2021-04-01	336	0	0	0	0.0	None	None	county	jhu-csse
0	02100	confirmed_incidence_num	2020-05-01	2021-04-01	335	0	0	0	0.0	None	None	county	jhu-csse
>>> df3
	geo_value	signal	time_value	issue	lag	missing_value	missing_stderr	missing_sample_size	value	stderr	sample_size	geo_type	data_source
0	02100	confirmed_7dav_incidence_num	2020-05-01	2021-04-01	335	0	0	0	0.0	None	None	county	jhu-csse
```

Track the lag values: the lag values of `confirmed_7dav_incidence_num` matches with the lag at the same date and time for the  `confirmed_cumulative_num` signal. 

This seems reasonable. A reasonable approach to computing the lag value for a derived value that depends on a window of previous values is this:
- take the window of values in the base signal (here, `confirmed_cumulative_num`) and find the largest issue (here, 2021-04-01)
- compute the distance between this largest issue and the time_value that we're computing for (generally the time_value we're computing for is the right-most point in the window, so in this case it's 2020-05-01), so here we get 335

I reimplemented this approach. Now I'm getting new issues with `min_lag`. Tracking it down, I found that the smoothed_incidence signal *just inherits the issue of the cumulative signal without taking a max over all the rows present in the summation*.

```python
>>> end_day = date(2020, 3, 12)
>>> df1 = covidcast.signal(data_source="jhu-csse", signal="confirmed_cumulative_num", start_day=end_day - timedelta(days=7), end_day=end_day, time_type="day", geo_type="state", geo_values="ak")
>>> df2 = covidcast.signal(data_source="jhu-csse", signal="confirmed_incidence_num", start_day=end_day - timedelta(days=6), end_day=end_day , time_type="day", geo_type="state", geo_values="ak")
>>> df3 = covidcast.signal(data_source="jhu-csse", signal="confirmed_7dav_incidence_num", start_day=end_day, end_day=end_day, time_type="day", geo_type="state", geo_values="ak")
>>> df1
geo_value	signal	time_value	issue	lag	missing_value	missing_stderr	missing_sample_size	value	stderr	sample_size	geo_type	data_source
0	ak	confirmed_cumulative_num	2020-03-05	2021-04-01	392	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_cumulative_num	2020-03-06	2021-04-01	391	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_cumulative_num	2020-03-07	2021-04-01	390	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_cumulative_num	2020-03-08	2021-04-01	389	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_cumulative_num	2020-03-09	2021-04-01	388	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_cumulative_num	2020-03-10	2021-04-01	387	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_cumulative_num	2020-03-11	2021-04-01	386	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_cumulative_num	2020-03-12	2020-10-29	231	0	0	0	0.0	None	None	state	jhu-csse

>>> df2
geo_value	signal	time_value	issue	lag	missing_value	missing_stderr	missing_sample_size	value	stderr	sample_size	geo_type	data_source
0	ak	confirmed_incidence_num	2020-03-06	2021-04-01	391	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_incidence_num	2020-03-07	2021-04-01	390	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_incidence_num	2020-03-08	2021-04-01	389	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_incidence_num	2020-03-09	2021-04-01	388	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_incidence_num	2020-03-10	2021-04-01	387	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_incidence_num	2020-03-11	2021-04-01	386	0	0	0	0.0	None	None	state	jhu-csse
0	ak	confirmed_incidence_num	2020-03-12	2020-10-29	231	0	0	0	0.0	None	None	state	jhu-csse

>>> df3
geo_value	signal	time_value	issue	lag	missing_value	missing_stderr	missing_sample_size	value	stderr	sample_size	geo_type	data_source
0	ak	confirmed_7dav_incidence_num	2020-03-12	2020-10-29	231	0	0	0	0.0	None	None	state	jhu-csse
```

**Comment by dshemetov** [comment]
Hopefully my notes above help understand my testing strategy. See the comments in test_meta-cache-updater

File: tests/acquisition/covidcast/test_covidcast_meta_cache_updater.py [hunk 1]
```diff
@@ -2,73 +2,234 @@
...
+            covidcast.signal(data_source="jhu-csse", signal="confirmed_cumulative_num", start_day=date(2020, 2, 13), end_day=end_day, time_type="day", geo_type="county", geo_values="02100,06003,06017,06027"),
+            covidcast.signal(data_source="jhu-csse", signal="confirmed_incidence_num", start_day=date(2020, 2, 14), end_day=end_day, time_type="day", geo_type="state", geo_values="ak,ca"),
+            covidcast.signal(data_source="jhu-csse", signal="confirmed_incidence_num", start_day=date(2020, 2, 14), end_day=end_day, time_type="day", geo_type="county", geo_values="02100,06003,06017,06027"),
+            covidcast.signal(data_source="jhu-csse", signal="confirmed_7dav_incidence_num", start_day=date(2020, 2, 20), end_day=end_day, time_type="day", geo_type="state", geo_values="ak,ca"),
+            covidcast.signal(data_source="jhu-csse", signal="confirmed_7dav_incidence_num", start_day=date(2020, 2, 20), end_day=end_day, time_type="day", geo_type="county", geo_values="02100,06003,06017,06027")
+        ])
+        >>> df.to_csv("test-data.csv")
+        """
+        self._insert_csv("repos/delphi/delphi-epidata/tests/acquisition/covidcast/test-data/test-data2.csv")
```
Comment (nolangormley) [inline tests/acquisition/covidcast/test_covidcast_meta_cache_updater.py:173]: There's a testdata folder in the delphi-epidata folder. It may allow for easier reuse of these flat files if needed for other tests.

Comment (nolangormley) [reply tests/acquisition/covidcast/test_covidcast_meta_cache_updater.py:173]: Another note on this. The directory structure here is assuming that the base working directory is above the repos folder. This is inconsistent with the t[est_utils.py](https://github.com/cmu-delphi/delphi-epidata/blob/dev/src/acquisition/covid_hosp/common/test_utils.py)

Comment (dshemetov) [reply tests/acquisition/covidcast/test_covidcast_meta_cache_updater.py:173]: Hm, I've never even looked at that file before. Looks like it's all for tests of `covid_hosp`, which I haven't really touched. 

Wait, it looks like `test_utils.py` does some complicated directory traversing up the tree until it can find `testdata`.

Comment (nolangormley) [reply tests/acquisition/covidcast/test_covidcast_meta_cache_updater.py:173]: Yeah you’re right. Let me dig through the source to see if that’s used elsewhere.

**Comment by dshemetov** [comment]
I'm going to merge this into `jit_computations` branch. With all the v4 updates to the tests, it duplicates the work to maintain keep these branches both in sync (and I spent a lot of time merging the v4 updates in today).

**Comment by dshemetov** [comment]
## A/B Testing Summary

I have completed a large batch of tests. Summary of findings:

1. JIT derived signals match exactly with Pandas-recomputed derived signals
2. JIT derived signals does not match exactly with live data, but in those cases
	- the live data contains bugs (e.g. possibly desync issues)
	- the live data contains artificial values (e.g. incidence earliest reference dates match cumulative, but it should be 1 day later)

(A **derived signal** is a signal obtained from a base signal via a transformation. A **base signal** is a signal that is not a derived signal.)

These findings give me confidence that we can roll this code out.

## Test Setup

### Data

I downloaded the latest data from:
- `sources=jhu-csse,usa-facts` 
- `signals=confirmed_cumulative_num,confirmed_incidence_num,confirmed_7dav_incidence_num`
- `geo_values=state:*,county:*` 
- `time_values=20200122-20220910`

### Test Procedure

The testing procedure went as follows:
1. I loaded all these signals into a local database
2. I obtained `confirmed_incidence_num` from JIT
	- I did this by querying the local API server using JIT (which computes the incidence values from `confirmed_cumulative_num` server-side)
3. I obtained `confirmed_incidence_num` from live data
	- I did this by querying the local API server *without* using JIT
4. I obtained `confirmed_incidence_num` by computing it from `confirmed_cumulative_num` with Pandas
	- I did this by reading the CSV file with `confirmed_cumulative_num` of all the USAFacts downloaded data into a dataframe, I reindexed it onto a contiguous time index, grouped by geo_value, and used [`diff`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.DataFrameGroupBy.diff.html?highlight=diff#pandas.core.groupby.DataFrameGroupBy.diff) to take differences

### Test Comparisons and Results

I joined these three series into a single dataframe and compared:
1. I searched for all the rows where Pandas and JIT were different in the `value` column; I found 0 rows where they were different!
2. I searched for all the rows where JIT and live data were different and broke these down into two cases
	1. JIT has a nan and live data doesn't; in all these cases JIT was right to have a nan (there was a nan in the 2-day window for incidence or a nan in the 7-day window for smoothed incidence)
	2. JIT and live data both have value; since JIT matched with Pandas and live data didn't, this is an error in the live data (possibly a cache desync)
